### PR TITLE
[SPARK-17699] Support for parsing JSON string columns

### PR DESCRIPTION
**Pull request by marmbrus**
Spark SQL has great support for reading text files that contain JSON data.  However, in many cases the JSON data is just one column amongst others.  This is particularly true when reading from sources such as Kafka.  This PR adds a new functions `from_json` that converts a string column into a nested `StructType` with a user specified schema.

Example usage:

``` scala
val df = Seq("""{"a": 1}""").toDS()
val schema = new StructType().add("a", IntegerType)

df.select(from_json($"value", schema) as 'json) // => [json: <a: int>]
```

This PR adds support for java, scala and python.  I leveraged our existing JSON parsing support by moving it into catalyst (so that we could define expressions using it).  I left SQL out for now, because I'm not sure how users would specify a schema.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 1]
```diff
@@ -467,3 +469,26 @@ case class JsonTuple(children: Seq[Expression])
   }
 }
 
+/**
+ * Converts an json input string to a [[StructType]] with the specified schema.
+ */
+case class JsonToStruct(schema: StructType, options: Map[String, String], child: Expression)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Should this override `ExpectsInputTypes`?

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Ah, yes, it definitly should.  Let me update.

**Comment by rxin** [comment]
Might want to send a dev list email to solicit feedback on the API?

**Comment by marmbrus** [comment]
Emailed the list.  Seems like a popular feature so far :)

**Comment by HyukjinKwon** [comment]
@marmbrus I just wonder if adding `to_json` make senses (although maybe it should be done in another PR). Just curious. I am just imaging the case to write out dataframes by some data sources not supporting nested structured types.

**Comment by yhuai** [comment]
LGTM. Merging to master.

**Comment by marmbrus** [comment]
@HyukjinKwon absolutely.  I actually changed the name from `json_parser` to `from_json` in anticipation of adding `to_json` :)

**Comment by DanielMe** [comment]
Is there any workaround I can use to achieve a similar effect in 1.6?

**Comment by yhuai** [comment]
@DanielMe The best options for 1.6 are `get_json_object ` and `json_tuple` (their docs can be found at https://spark.apache.org/docs/1.6.0/api/scala/index.html#org.apache.spark.sql.functions$).

**Comment by DanielMe** [comment]
@yhuai thanks! My impression was that `get_json_object` does not convert json arrays to `ArrayType`s, maybe I misunderstood the way it's supposed to be used though.

**Comment by yhuai** [comment]
@DanielMe oh, I see. `get_json_object` will not parse json array. You need to have a UDF to do that for Spark 1.6.

**Comment by gatorsmile** [comment]
Actually, to specify the schema in SQL language, maybe we can use a JSON string. A little bit odd. Let us see whether users need it in SQL

**Comment by gatorsmile** [comment]
Based on the comment @marmbrus in a JIRA, we prefer to using our DDL format. For example, like what we did for CREATE TABLE, we can specify the schema using `a int, b string`

**Pull request closed without merging**